File: force-field/dart/particle.dart
import 'vector.dart';

class Particle {
  Vector position;
  double mass;
  Vector velocity;
  
  Particle(this.position, this.mass) {
    velocity = new Vector(0, 0);
  }
  
  Vector forceFrom(Particle p) {
    Vector distance = p.position - position;
    double magnitude = mass * p.mass / distance.dot(distance);
    return distance.normalize() * magnitude;
  }
  
  void applyForce(Vector force, double timeDelta) {
    velocity.addInPlace(force * (timeDelta / mass.abs()));
    position.addInPlace(velocity * timeDelta);
  }
  
  String toString() {
    return "{Particle mass:${mass} velocity: ${velocity}" +
      " position: ${position}}";
  }

}

File: force-field/php/Particle.php
<?php

class Particle {
	public $position;
	public $mass;
	public $velocity;

	public function __construct($pos, $mass)
	{
		$this->position = $pos;
		$this->mass = $mass;
		$this->velocity = new Vector(0, 0);
	}

	public function forceFrom(Particle $p)
	{
		$distance = $p->position->subtract($this->position);
		$magnitude = $this->mass * $p->mass / $distance->dot($distance);
		return $distance->normalize()->scale($magnitude);
	}

	public function applyForce($force, $timeDelta)
	{
		$this->velocity->addInPlace($force->scale($timeDelta / abs($this->mass)));
		$this->position->addInPlace($this->velocity->scale($timeDelta));
	}

	public function __toString()
	{
		return sprintf('{Particle mass: %d velocity: %s position: %s', $this->mass, $this->velocity, $this->position);
	}
}
